File: backend/app/services/categorizer.py
from __future__ import annotations

import re

from sqlalchemy.orm import Session

from app.models.category_rule import CategoryRule
from app.models.transaction import Transaction


class Categorizer:
    """Rule-based transaction categorizer.

    Loads all active CategoryRule rows ordered by priority ascending.
    For each transaction description, evaluates rules in order. First match wins.
    """

    def __init__(self, db: Session) -> None:
        self._rules: list[CategoryRule] = []
        self._load_rules(db)

    def _load_rules(self, db: Session) -> None:
        """Load all active rules ordered by priority ascending."""
        self._rules = (
            db.query(CategoryRule)
            .filter(CategoryRule.is_active.is_(True))
            .order_by(CategoryRule.priority.asc())
            .all()
        )

    def categorize(self, description: str) -> int | None:
        """Run description against rules in priority order.

        Returns category_id of first match, or None.
        - substring: case-insensitive containment check
        - regex: re.search with IGNORECASE
        """
        for rule in self._rules:
            if rule.match_type == "substring":
                if rule.pattern.upper() in description.upper():
                    return rule.category_id
            elif rule.match_type == "regex":
                if re.search(rule.pattern, description, re.IGNORECASE):
                    return rule.category_id
        return None

    def recategorize_all(self, db: Session) -> int:
        """Re-run all rules against all non-manually-categorized transactions.

        Returns the count of updated rows.
        """
        # Reload rules to pick up any changes
        self._load_rules(db)

        transactions = (
            db.query(Transaction).filter(Transaction.is_manually_categorized.is_(False)).all()
        )

        updated_count = 0
        for txn in transactions:
            new_category_id = self.categorize(txn.description)
            if new_category_id != txn.category_id:
                txn.category_id = new_category_id
                updated_count += 1

        db.commit()
        return updated_count

File: frontend/src/lib/local-engine/categorizer.ts
import type { BudgetCategoryRule } from "./types";

export function categorize(
  description: string,
  rules: BudgetCategoryRule[]
): number | null {
  const activeRules = rules
    .filter((r) => r.is_active)
    .sort((a, b) => a.priority - b.priority);

  for (const rule of activeRules) {
    if (rule.match_type === "substring") {
      if (description.toUpperCase().includes(rule.pattern.toUpperCase())) {
        return rule.category_id;
      }
    } else if (rule.match_type === "regex") {
      try {
        if (new RegExp(rule.pattern, "i").test(description)) {
          return rule.category_id;
        }
      } catch {
        // Invalid regex — skip
      }
    }
  }
  return null;
}

export function recategorizeAll(store: {
  transactions: { description: string; category_id: number | null; is_manually_categorized: boolean }[];
  categoryRules: BudgetCategoryRule[];
}): number {
  let updated = 0;
  for (const txn of store.transactions) {
    if (txn.is_manually_categorized) continue;
    const newCatId = categorize(txn.description, store.categoryRules);
    if (newCatId !== txn.category_id) {
      txn.category_id = newCatId;
      updated++;
    }
  }
  return updated;
}
